src/main/java/net/helix/pendulum/controllers/RoundViewModel.java

```diff
@@ -372,6 +367,44 @@ public static int getRoundIndex(TransactionViewModel milestoneTransaction) {
         return tips;
     }
 
+    public Set<Hash> getConfirmedTransactions(Tangle tangle, int security) throws Exception {
+
+        Set<Hash> transactions = new HashSet<>();
+        Set<Hash> tips = getConfirmedTips(tangle, security);
+
+        final Queue<Hash> nonAnalyzedTransactions = new LinkedList<>(tips);
+        Hash hashPointer;
+        while ((hashPointer = nonAnalyzedTransactions.poll()) != null) {
+            final TransactionViewModel transaction = fromHash(tangle, hashPointer);
+            if (transaction.snapshotIndex() == index()) {
+                transactions.add(hashPointer);
+                // transaction of milestone bundle
+                TransactionViewModel milestoneTx;
+                if ((milestoneTx = transaction.isMilestoneBundle(tangle)) != null) {
+                    Set<Hash> parents = RoundViewModel.getMilestoneTrunk(tangle, transaction, milestoneTx);
+                    parents.addAll(RoundViewModel.getMilestoneBranch(tangle, transaction, milestoneTx, security));
+                    for (Hash parent : parents) {
+                        nonAnalyzedTransactions.offer(parent);
+                    }
+                }
+                // normal transaction
+                else {
+                    nonAnalyzedTransactions.offer(transaction.getTrunkTransactionHash());
+                    nonAnalyzedTransactions.offer(transaction.getBranchTransactionHash());
+                }
+            } else {
+                break;
+            }
+        }
+        return transactions;
+    }
+
+    public boolean isTransactionConfirmed(Tangle tangle, int security, Hash transaction) throws Exception {
+        Set<Hash> confirmedTransactions = getConfirmedTransactions(tangle, security);
```

Comment:
Do we  traverse the tangle each time `isTransactionConfirmed()` is called? Can we save the confirmation status in the `TransactionViewModel` of all `confirmedTransactions` ?

Reply:
`isTransactionConfirmed` is not used anymore. it is obsolete!

Reply:
The idea is to count here: https://github.com/HelixNetwork/pendulum/blob/confirmation-state/src/main/java/net/helix/pendulum/service/milestone/impl/MilestoneTrackerImpl.java#L301 
using: https://github.com/HelixNetwork/pendulum/blob/confirmation-state/src/main/java/net/helix/pendulum/service/milestone/impl/MilestoneTrackerImpl.java#L303: `getReferencedTransactions`